File: SEM 3/PCPF/Experiments/EXP10/TicTacToe.hs
import Data.List (intercalate, transpose,intersperse)

type Board = [[Char]]
type Position = (Int, Int)

-- Initialize an empty board
initBoard :: Board
initBoard = replicate 3 (replicate 3 ' ')

-- Print the board
printBoard :: Board -> IO ()
printBoard board = putStrLn $ intercalate "\n-----\n" (map (intersperse '|') board)

-- Check if there's a winner
checkWinner :: Board -> Maybe Char
checkWinner board = 
    let rows = board
        cols = transpose board
        diags = [[board !! i !! i | i <- [0..2]], [board !! i !! (2-i) | i <- [0..2]]]
        lines = rows ++ cols ++ diags
    in  case filter (\line -> all (== 'X') line) lines of
            (_:_) -> Just 'X'
            [] -> case filter (\line -> all (== 'O') line) lines of
                (_:_) -> Just 'O'
                [] -> Nothing

-- Make a move
makeMove :: Board -> Position -> Char -> Board
makeMove board (row, col) player = 
    let (r1, r2:r3) = splitAt row board
        (c1, _:c2) = splitAt col r2
    in r1 ++ [c1 ++ [player] ++ c2] ++ r3

-- Check if the board is full
isFull :: Board -> Bool
isFull = all (all (/= ' '))

-- Parse input safely
parseInput :: String -> Maybe Position
parseInput input =
    case reads input :: [((Int, Int), String)] of
        [((x, y), "")] -> Just (x, y)
        _              -> Nothing

-- Main game loop
playGame :: Board -> Char -> IO ()
playGame board player = do
    printBoard board
    case checkWinner board of
        Just winner -> putStrLn $ "Player " ++ [winner] ++ " wins!"
        Nothing -> 
            if isFull board
            then putStrLn "It's a draw!"
            else do
                putStrLn $ "Player " ++ [player] ++ ", enter row and column (format: x,y): "
                input <- getLine
                case parseInput input of
                    Just (row, col) | row >= 0 && row < 3 && col >= 0 && col < 3 && (board !! row !! col) == ' ' -> 
                        playGame (makeMove board (row, col) player) (if player == 'X' then 'O' else 'X')
                    _ -> do
                        putStrLn "Invalid move, try again."
                        playGame board player

main :: IO ()
main = playGame initBoard 'X'

File: SEM 3/PCPF/Experiments/EXP10/TicTacToe.java
package EXP10;

import java.util.Scanner;

public class TicTacToe {
    public static char checkWinner(char[][] board) {
        // Check rows and columns
        for (int i = 0; i < 3; i++) {
            if (board[i][0] == board[i][1] && board[i][1] == board[i][2] && board[i][0] != ' ')
                return board[i][0];
            if (board[0][i] == board[1][i] && board[1][i] == board[2][i] && board[0][i] != ' ')
                return board[0][i];
        }

        // Check diagonals
        if (board[0][0] == board[1][1] && board[1][1] == board[2][2] && board[0][0] != ' ')
            return board[0][0];
        if (board[0][2] == board[1][1] && board[1][1] == board[2][0] && board[0][2] != ' ')
            return board[0][2];

        return ' ';
    }

    public static void printBoard(char[][] board) {
        for (char[] row : board) {
            for (char cell : row) {
                System.out.print((cell == ' ' ? '-' : cell) + " ");
            }
            System.out.println();
        }
    }

    public static void main(String[] args) {
        char[][] board = new char[3][3];
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                board[i][j] = ' ';
            }
        }

        char winner = ' ';
        int moves = 0;
        char currentPlayer = 'X';
        Scanner scanner = new Scanner(System.in);

        while (winner == ' ' && moves < 9) {
            printBoard(board);
            System.out.print("Player " + currentPlayer + ", enter row and column (0-2): ");
            int row = scanner.nextInt();
            int col = scanner.nextInt();

            if (row >= 0 && row < 3 && col >= 0 && col < 3 && board[row][col] == ' ') {
                board[row][col] = currentPlayer;
                winner = checkWinner(board);
                currentPlayer = (currentPlayer == 'X') ? 'O' : 'X';
                moves++;
            } else {
                System.out.println("Invalid move, try again.");
            }
        }

        printBoard(board);
        if (winner != ' ')
            System.out.println("Player " + winner + " wins!");
        else
            System.out.println("It's a draw!");

        scanner.close();
    }
}
